### DISCUSSION TITLE
Efficient and Differentiable Degree Calculation on Vectorized Adjacency Matrices

### DISCUSSION THREAD

**Discussion by maxwass** [jax-ml/jax#16166] [Q&A]
### Problem Overview
I'm working with a batch of symmetric adjacency matrices, `A`, that have a shape of `[batch_size, n, n]`. These graphs are weighted and may not be sparse. For my application, I have to work with a vectorized form of these adjacencies by taking the upper triangular part of each slice of `A`, thus creating `A_vec` with a shape of `[batch_size, n*(n-1)/2]`.

### Objective
My task is to calculate the degrees of each node in `A_vec`. To do this, I can define a binary array `D \in {0, 1}^{n \times n(n-1)/2}` so that `A1 = D*A_vec`. The `D` matrix is just a fancy indexing matrix, with its j-th row indicating which indices in `A_vec` represent edges adjacent to node j. As `n` increases, `D` will become very sparse with `n(n-1)/2` nonzero entries out of a total of `n^2(n-1)` entries.

The raw matrix multiplication `D*A_vec` is currently a limiting factor in my application as I am performing iterative optimization to refine `A_vec`. This operation needs to be run many times within the inner step of a `fori_loop`, which can run from 200 to 5000 iterations. Therefore, I need it to be both jit compilable and differentiable.

### Attempts and Issues Faced
- I tried to make `D` sparse using `sparse.BCOO.fromdense(D)` and perform the loop in raw python. While this works, it is extremely slow.
- If I try to use my looping function (with `@jit` and `@sparsify` decorators) with the `fori_loop`, then I get an error (see below) during the matmul operation.
- I'm definitely open to simply indexing `A_vec` and summing, e.g. `A_vec[adjecent_edges_indices].sum(axis=-1)` to recover node degrees, but I couldn't figure out a way to do this in Jax while retaining Jit compatibility and differentiability.

### Error Message
File.../lib/python3.9/site-packages/jax/experimental/sparse/transform.py:462), in _sparsify_with_interpreter..wrapped(*args, **params)
    460 spenv = SparsifyEnv()
    461 spvalues = arrays_to_spvalues(spenv, args)
--> 462 spvalues_out, out_tree = f_raw(spenv, *spvalues, **params)
    463 out = spvalues_to_arrays(spenv, spvalues_out)
    464 return tree_unflatten(out_tree, out)

File ..., in sparsify_raw..wrapped(spenv, *spvalues, **params)
    445 in_avals_flat = spvalues_to_avals(spenv, spvalues_flat)
    446 wrapped_fun, out_tree = flatten_fun_nokwargs(lu.wrap_init(f, params), in_tree)
...
    341                 if not _arraylike(arg))
    342 msg = "{} requires ndarray or scalar arguments, got {} at position {}."
--> 343 raise TypeError(msg.format(fun_name, type(arg), pos))

TypeError: matmul requires ndarray or scalar arguments, got  at position 1.

**Comment by jakevdp** [comment]
Sparse operations can be useful for saving memory compared to dense operations, but in general unless your matrices are very sparse (say 99% or 99.9% sparsity depending on the matmul implementation) unstructured sparse operations are not going to lead to faster runtime than equivalent dense operations. So with that in mind, I'd probably look to other means of optimizing your operation. With only a verbal description of your code, though, I'm not sure what to suggest.

**Comment by maxwass** [comment]
Apologies, I thought the code may act to obfuscate the real question.

Below I have included code for the construction of D (included in case it may be useful), the inner `step`, and outer `looping` function. I believe the operations involving D dominate both the runtime and memory used by the program; both runtime and memory are important for me to optimize as I'm interested in (repeatedly) running this on large inputs (i.e. large `n`).

Alpha, beta, and gamma are learnable parameters. I would like the resulting `looping` function to be differentiable w.r.t. all learnable parameters and input `x`.

```
def construct_D(n: int):
    """
        Returns a linear transformation D such that Dw = A1, where A is the adjacency matrix of an undirected graph
        without self loops and w is the upper triangular part of A (without diagonal).

        D is of shape 'n x (n(n-1)/2)'.

    """
    dtype = np.float32

    # total number of possible edges in undirected graph without self loops n(n-1)/2
    NumCol = int((n * (n - 1) / 2))
    I = np.zeros(NumCol, dtype=dtype)
    J = np.zeros(NumCol, dtype=dtype)
    k = 0
    for i in np.arange(n - 1):
        I[k:k + n - i - 1] = np.arange(i + 1, n)
        k = k + n - i - 1
    k = 0
    for i in np.arange(n - 1):
        J[k:k + n - i - 1] = i
        k = k + n - i - 1
    Jidx = np.concatenate((I, J), axis=0)
    Iidx = np.tile(np.arange(NumCol, dtype=dtype), 2)

    Jidx, Iidx = Jidx.astype(np.int32), Iidx.astype(np.int32)
    Dt = np.zeros((NumCol, n), dtype=dtype)
    Dt[Iidx, Jidx] = 1
    D = Dt.T
    return D

def step(
        x: jnp.ndarray,
        w_prev: jnp.ndarray,
        v_prev: jnp.ndarray,
        alpha: float,
        beta: float,
        gamma: float,
        D: jnp.ndarray,
        eps: float = 1e-16
):
    y1 = w_prev - gamma * (2 * beta * w_prev + 2 * x + jnp.matmul(v_prev, D))
    y2 = v_prev + gamma * jnp.matmul(w_prev, D.T)

    p1 = jax.nn.relu(y1)
    up = y2 ** 2 + 4 * gamma * alpha
    up = jnp.clip(up, a_min=eps)

    p2 = (y2 - jnp.sqrt(up)) / 2

    q1 = p1 - gamma * (2 * beta * p1 + 2 * x + jnp.matmul(p2, D)) # p2.matmul(D))
    q2 = p2 + gamma * jnp.matmul(p1, D.T)

    w = w_prev - y1 + q1
    v = v_prev - y2 + q2

    return w, v

@sparsify
@partial(jit, static_argnames=['num_steps'])
def looping(
        x: jnp.ndarray,
        w_init: jnp.ndarray,
        v_init: jnp.ndarray,
        alpha: jnp.ndarray,
        beta: jnp.ndarray,
        gamma: jnp.ndarray,
        num_steps: int,
        D: jnp.ndarray,
):
    """
    Arguments:
        x is the input data of shape [batch_size, n*(n-1)//2]
        w_init, v_init are the initial values of the optimization parameters, with 
             shapes [batch_size, n*(n-1)//2] and [batch_size, n], respectively.
        alpha, beta, and gamma are the learnable parameters. Shape [num_steps]
        num_steps is the total number of iterations to run the loop
        D is linear transform mapping vectorized adjacency into degree vector
    """
    f = partial(step, D=D, x=x)

    def body_fun(i, carry):
        w, lam = carry
        w, lam = f(w_prev=w, v_prev=lam, alpha=alpha[i], beta=beta[i], gamma=gamma[i])
        return w, lam

    w, v = jax.lax.fori_loop(0, num_steps, body_fun, (w_init, v_init))
    return w, v

# sample execution
n, batch_size, num_steps = 100, 50, 4000
x = jax.random.uniform(subkey, (batch_size, n*(n-1)//2)) # sample inputs
alpha, beta, gamma = 0.5*jpn.ones(num_steps), 0.5*jnp.ones(num_steps), 1.0*jnp.ones(num_steps)
D = sparse.BCOO.fromdense(construct_D(n))
w_init, v_init = jnp.ones(batch_size, n*(n-1)//2), jnp.ones(batch_size, n*(n-1)//2)
w, v = looping(x, w_init, v_init, alpha, beta, gamma, num_steps, D) # This fails with the matmul error when using sparse D
```

**Reply by jakevdp** [reply]
What is `subkey` in this code?

**Reply by jakevdp** [reply]
OK, I used `subkey = jax.random.PRNGKey(0)`, and found that there's a shape error in your matmul, whether `D` is sparse or dense.

Can you double-check the code?

**Reply by maxwass** [reply]
Ah sorry, there is an error in the shape of v_init, should be `jnp.ones(batch_size, n)`. The code below is now copy-pastable, it is the same as the code in the Colab link included.

Here is some more information on the errors I see get when tinkering with @jit and @sparsify decorators

_Local Mac-M2 Machine_
use_sparse = False,  @jit on looping() --> works fine (but not scalable)
use_sparse = True, @jit on looping() --> `ValueError: Non-hashable static arguments are not supported, as this can lead to unexpected cache-misses. Static argument (index 6) of type  for function looping is non-hashable.`
use_sparse = True, no @jit on looping() --> `TypeError: bind() got an unexpected keyword argument 'call_jaxpr'`

_[Google Colab](https://colab.research.google.com/drive/152rrbWZ6nbYefZzQBaRZBCqDB-AWydSy?usp=sharing)_. Runtime was restarted before each case.
use_sparse = False,  @jit on looping() --> works fine (but not scalable)
use_sparse = True, @jit on looping() --> `TypeError: CustomJVPCallPrimitive.bind() got an unexpected keyword argument 'call_jaxpr'`
use_sparse = True, no @jit on looping() --> `TypeError: CustomJVPCallPrimitive.bind() got an unexpected keyword argument 'call_jaxpr`


My goal is to scale this to very large inputs `n`. This function will be called many times (and so I want it to be compilable). I want to run as fast a possible while also optimizing memory, as currently the memory usage scales as `n^3` from the `D` term (which again, is just a fancy index function).


```
import numpy as np
from functools import partial
import jax
from jax import jit
import jax.numpy as jnp
from jax.experimental.sparse import sparsify
import jax.experimental.sparse as sparse


def construct_D(n: int):
    """
        Returns a linear transformation D such that Dw = A1, where A is the adjacency matrix of an undirected graph
        without self loops and w is the upper triangular part of A (without diagonal).

        D is of shape n x (n(n-1)/2)

    """
    dtype = np.float32

    # total number of possible edges in undirected graph without self loops n(n-1)/2
    NumCol = int((n * (n - 1) / 2))
    I = np.zeros(NumCol, dtype=dtype)
    J = np.zeros(NumCol, dtype=dtype)
    k = 0
    for i in np.arange(n - 1):
        I[k:k + n - i - 1] = np.arange(i + 1, n)
        k = k + n - i - 1
    k = 0
    for i in np.arange(n - 1):
        J[k:k + n - i - 1] = i
        k = k + n - i - 1
    Jidx = np.concatenate((I, J), axis=0)
    Iidx = np.tile(np.arange(NumCol, dtype=dtype), 2)

    Jidx, Iidx = Jidx.astype(np.int32), Iidx.astype(np.int32)
    Dt = np.zeros((NumCol, n), dtype=dtype)
    Dt[Iidx, Jidx] = 1
    D = Dt.T
    return D

def step(
        x: jnp.ndarray,
        w_prev: jnp.ndarray,
        v_prev: jnp.ndarray,
        alpha: float,
        beta: float,
        gamma: float,
        D: jnp.ndarray,
        eps: float = 1e-16
):
    y1 = w_prev - gamma * (2 * beta * w_prev + 2 * x + jnp.matmul(v_prev, D))
    y2 = v_prev + gamma * jnp.matmul(w_prev, D.T)

    p1 = jax.nn.relu(y1)
    up = y2 ** 2 + 4 * gamma * alpha
    up = jnp.clip(up, a_min=eps)

    p2 = (y2 - jnp.sqrt(up)) / 2

    q1 = p1 - gamma * (2 * beta * p1 + 2 * x + jnp.matmul(p2, D)) # p2.matmul(D))
    q2 = p2 + gamma * jnp.matmul(p1, D.T)

    w = w_prev - y1 + q1
    v = v_prev - y2 + q2

    return w, v

@partial(jit, static_argnames=['num_steps'])
def looping(
        x: jnp.ndarray,
        w_init: jnp.ndarray,
        v_init: jnp.ndarray,
        alpha: jnp.ndarray,
        beta: jnp.ndarray,
        gamma: jnp.ndarray,
        num_steps: int,
        D: jnp.ndarray,
):
    """
    Arguments:
        x is the input data of shape [batch_size, n*(n-1)//2]
        w_init, v_init are the initial values of the optimization parameters, with 
             shapes [batch_size, n*(n-1)//2] and [batch_size, n], respectively.
        alpha, beta, and gamma are the learnable parameters. Shape [num_steps]
        num_steps is the total number of iterations to run the loop
        D is linear transform mapping vectorized adjacency into degree vector
    """
    f = partial(step, x=x, D=D)

    def body_fun(i, carry):
        w, v = carry
        w, v = f(w_prev=w, v_prev=v, alpha=alpha[i], beta=beta[i], gamma=gamma[i])
        return w, v

    w, v = jax.lax.fori_loop(0, num_steps, body_fun, (w_init, v_init))
    return w, v


n, batch_size, num_steps = 20, 50, 4000

use_sparse = False
looping = sparsify(looping) if use_sparse else looping
D = sparse.BCOO.fromdense(construct_D(n)) if use_sparse else jnp.array(construct_D(n))

# sample data, initialize parameters and optimization variables, then execute the loop
x = jax.random.uniform(jax.random.PRNGKey(0), shape=(batch_size, n*(n-1)//2)) 
alpha, beta, gamma = 0.5*jnp.ones(num_steps), 0.5*jnp.ones(num_steps), 1.0*jnp.ones(num_steps)
w_init, v_init = jnp.ones((batch_size, n*(n-1)//2)), jnp.ones((batch_size, n))

w, v = looping(x, w_init, v_init, alpha, beta, gamma, num_steps, D)
```

**Reply by jakevdp** [reply]
Thanks - with the most recent version of jax, the sparsify transformation seems to be crashing on `custom_jvp_call`. We'd need to update the transformation to handle this.

**Reply by maxwass** [reply]
Ok, so it seem's like sparsify will not be usable for this in the near future.

Are there any other significant optimizations you can suggest here for runtime and/or memory?
